File: problems/0100-0200/0200_number-of-islands/solution_bfs.go
func numIslands(grid [][]string) int {
	rows := len(grid)
	if rows == 0 {
		return 0
	}
	cols := len(grid[0])
	visited := make([][]bool, rows)
	for i := range visited {
		visited[i] = make([]bool, cols)
	}
	count := 0
	dirs := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == "1" && !visited[r][c] {
				count++
				queue := [][2]int{{r, c}}
				visited[r][c] = true
				for head := 0; head < len(queue); head++ {
					x, y := queue[head][0], queue[head][1]
					for _, d := range dirs {
						nx, ny := x+d[0], y+d[1]
						if nx >= 0 && nx < rows && ny >= 0 && ny < cols &&
							grid[nx][ny] == "1" && !visited[nx][ny] {
							visited[nx][ny] = true
							queue = append(queue, [2]int{nx, ny})
						}
					}
				}
			}
		}
	}
	return count
}

File: problems-adapt/0101-0200/0200_count-grid-islands/solution_bfs.rs
use std::collections::VecDeque;

impl Solution {
    pub fn count_grid_islands(grid: Vec<Vec<String>>) -> i32 {
        let rows = grid.len();
        if rows == 0 {
            return 0;
        }
        let cols = grid[0].len();
        let mut visited = vec![vec![false; cols]; rows];
        let mut count = 0;
        let dirs = [(1i32, 0i32), (-1, 0), (0, 1), (0, -1)];
        for r in 0..rows {
            for c in 0..cols {
                if grid[r][c] == "1" && !visited[r][c] {
                    count += 1;
                    let mut queue: VecDeque<(usize, usize)> = VecDeque::new();
                    queue.push_back((r, c));
                    visited[r][c] = true;
                    while let Some((x, y)) = queue.pop_front() {
                        for (dx, dy) in dirs {
                            let nx = x as i32 + dx;
                            let ny = y as i32 + dy;
                            if nx >= 0 && ny >= 0 && (nx as usize) < rows && (ny as usize) < cols {
                                let (nx, ny) = (nx as usize, ny as usize);
                                if grid[nx][ny] == "1" && !visited[nx][ny] {
                                    visited[nx][ny] = true;
                                    queue.push_back((nx, ny));
                                }
                            }
                        }
                    }
                }
            }
        }
        count
    }
}
